doc/design/logging.md

```diff
@@ -0,0 +1,396 @@
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+
+int main()
+{
+    iox::log::Logger::init(iox::log::logLevelFromEnvOr(iox::log::LogLevel::DEBUG));
+
+    IOX_LOG(DEBUG) << "Hello World";
+
+    return 0;
+}
+```
+
+#### Add logging support for custom types
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+#include "iceoryx_hoofs/log/logstream.hpp"
+#include <cstdint>
+
+struct MyType
+{
+    uint16_t a {0};
+    uint16_t b {0};
+};
+
+iox::log::LogStream& operator<<(iox::log::LogStream& stream, const MyType& m)
+{
+    stream << "MyType { a = " << m.a << "; b = " << m.b << " }";
+    return stream;
+}
+
+int main()
+{
+    MyType m;
+    IOX_LOG(INFO) << m;
+
+    return 0;
+}
+```
+
+#### Creating a custom logger
+
+This is a simple custom logger which prints the log level as emoji and ignores
+the file, line number and function name.
+
+```cpp
+#include "iceoryx_hoofs/log/logger.hpp"
+#include "iceoryx_hoofs/log/logging.hpp"
+
+class MyLogger : public iox::log::Logger
+{
+  public:
+    static void init()
+    {
+        static MyLogger myLogger;
+        iox::log::setActiveLogger(&myLogger);
+        iox::log::Logger::init(iox::log::logLevelFromEnvOr(iox::log::LogLevel::INFO));
+    }
+
+  private:
+    void setupNewLogMessage(const char*, const int, const char*, iox::log::LogLevel logLevel) override
+    {
+        switch(logLevel) {
+            case iox::log::LogLevel::FATAL:
+                logString("💀: ");
+                break;
+            case iox::log::LogLevel::ERROR:
+                logString("🙈: ");
+                break;
+            case iox::log::LogLevel::WARN:
+                logString("🙀: ");
+                break;
+            case iox::log::LogLevel::INFO:
+                logString("💘: ");
+                break;
+            case iox::log::LogLevel::DEBUG:
+                logString("🐞: ");
+                break;
+            case iox::log::LogLevel::TRACE:
+                logString("🐾: ");
+                break;
+            default:
+                logString("🐔: ");
+        }
+    }
+
+    void flush() override {
+        puts(iox::log::Logger::buffer());
+        iox::log::Logger::assumePushed();
+    }
+};
+
+int main()
+{
+    MyLogger::init();
+
+    IOX_LOG(FATAL) << "Whoops ... look, over there is a dead seagull flying!";
+    IOX_LOG(ERROR) << "Oh no!";
+    IOX_LOG(WARN) << "It didn't happen!";
+    IOX_LOG(INFO) << "All glory to the hypnotoad!";
+    IOX_LOG(DEBUG) << "I didn't do it!";
+    IOX_LOG(TRACE) << "Row row row your boat!";
+
+    return 0;
+}
+```
+
+## Open issues
+
+- do we need to change the log level after `Logger::init`
+- do we want a `IOX_LOG_IF(cond, level)` macro
+- shall the TestingLogger register signals to catch SIGTERM, etc. and print the
+  log messages when the signal is raised? It might be necessary to wait for the
+  error handling refactoring before this can be done
+- instead of having the `Logger::init()` static function with hidden default
```

Comment:
Does it make sense to enforce initialization of the logger by design without ever having to call `init`?

This can be done, but it cannot be done automatically as the very first static initialization I think. To do enforce this, a static dummy object defined in the executable first which calls `init` in its constructor may be sufficient (not 100% sure, need to research this).

The benefit is that then it can be used in any static, stack or heap object allocated afterwards.